File: src/gaia/daemon/lock.py
# Copyright(C) 2025-2026 Advanced Micro Devices, Inc. All rights reserved.
# SPDX-License-Identifier: MIT
"""Cross-platform advisory file lock serializing daemon start.

Only ONE process may be in the "decide + spawn" critical section at a time, so two
concurrent ``start_or_attach()`` callers spawn exactly one daemon (the loser waits,
then attaches to the winner's instance.json). An OS advisory lock is used rather
than an ``O_EXCL`` create-lock because the OS releases it automatically when the
holder dies — a create-lock would strand a stale lock file after SIGKILL.

POSIX uses ``fcntl.flock``; Windows uses ``msvcrt.locking``. Acquisition is a
bounded non-blocking retry so an abandoned attempt fails loudly with an actionable
error instead of hanging forever.
"""

from __future__ import annotations

import os
import time
from typing import Optional

from gaia.daemon import paths
from gaia.daemon.errors import DaemonLockError

if os.name == "nt":
    import msvcrt
else:
    import fcntl


class StartLock:
    """Context manager holding the exclusive daemon-start lock."""

    def __init__(self, timeout: float = 30.0, poll: float = 0.1):
        self._timeout = timeout
        self._poll = poll
        self._fd: Optional[int] = None
        self._path = paths.lock_path()

    def _try_acquire(self) -> bool:
        if os.name == "nt":
            try:
                msvcrt.locking(self._fd, msvcrt.LK_NBLCK, 1)
                return True
            except OSError:
                return False
        try:
            fcntl.flock(self._fd, fcntl.LOCK_EX | fcntl.LOCK_NB)
            return True
        except OSError:
            return False

    def __enter__(self) -> "StartLock":
        paths.ensure_host_dir()
        # 0600: the lock file lives beside the token-bearing instance.json.
        self._fd = os.open(str(self._path), os.O_RDWR | os.O_CREAT, 0o600)
        deadline = time.monotonic() + self._timeout
        while True:
            if self._try_acquire():
                return self
            if time.monotonic() >= deadline:
                os.close(self._fd)
                self._fd = None
                raise DaemonLockError(
                    f"could not acquire the daemon start lock at {self._path} within "
                    f"{self._timeout}s — another process is starting the daemon and "
                    "did not finish. Check for a stuck `gaia daemon`/UI start, then "
                    "retry; if it persists inspect the daemon log via `gaia daemon logs`."
                )
            time.sleep(self._poll)

    def __exit__(self, exc_type, exc, tb) -> None:
        if self._fd is None:
            return
        try:
            if os.name == "nt":
                try:
                    msvcrt.locking(self._fd, msvcrt.LK_UNLCK, 1)
                except OSError:
                    pass
            else:
                fcntl.flock(self._fd, fcntl.LOCK_UN)
        finally:
            os.close(self._fd)
            self._fd = None

File: tui/internal/daemon/lock.go
package daemon

import (
	"fmt"
	"os"
	"time"
)

// lockPoll is the retry interval while waiting for the start lock.
const lockPoll = 100 * time.Millisecond

// fileLock holds the exclusive daemon-start lock. Only ONE process may be in the
// "decide + spawn" critical section at a time, so two concurrent StartOrAttach
// callers yield exactly one daemon (the loser attaches to the winner's
// instance.json).
//
// An OS advisory lock is used rather than an O_EXCL create-lock because the OS
// releases it when the holder dies — a create-lock would strand a stale lock
// file after SIGKILL.
type fileLock struct {
	f *os.File
}

// acquireLock takes the exclusive lock at path, retrying until timeout. Failure
// is loud: an abandoned start attempt must surface an actionable error rather
// than hang forever.
func acquireLock(path string, timeout time.Duration) (*fileLock, error) {
	// 0600: the lock file lives beside the token-bearing instance.json.
	f, err := os.OpenFile(path, os.O_RDWR|os.O_CREATE, 0o600)
	if err != nil {
		return nil, &StartError{Reason: fmt.Sprintf("cannot open the start lock at %s: %v", path, err)}
	}
	deadline := time.Now().Add(timeout)
	for {
		ok, lerr := tryLock(f)
		if lerr != nil {
			f.Close()
			return nil, &StartError{Reason: fmt.Sprintf("cannot lock %s: %v", path, lerr)}
		}
		if ok {
			return &fileLock{f: f}, nil
		}
		if time.Now().After(deadline) {
			f.Close()
			return nil, &StartError{Reason: fmt.Sprintf(
				"another process held the start lock at %s for more than %s without finishing — "+
					"look for a stuck `gaia daemon start` or Agent UI launch", path, timeout)}
		}
		time.Sleep(lockPoll)
	}
}

func (l *fileLock) release() {
	if l == nil || l.f == nil {
		return
	}
	unlock(l.f)
	l.f.Close()
	l.f = nil
}
